Fix for issue #875: Custom translations for one language showing for other languages

BrDewit opened the pull request:
In issue #875 I outlined an issue where creating a custom translation for one language, caused all other languages to display that custom translation, instead of only the specified language displaying the custom translation.

The cause of the issue was due to the fact that the code fell back on the custom translations of other languages when creating the translation function of a specific language. If there was no custom translation for the language at all, for any key, the custom translation of either the default language, or if that didn't exist, the first specified custom translation, was used.
If any custom translation existed for a certain language, the issue would not occur as the `messagesByLanguageTag_themeDefined[currentLanguage.languageTag]` call on line 223 of the getI18n.tsx file does not return undefined in this case, and it would not fall back on another language.

I've removed the fallback to other languages in this part of the code, as the custom translation is supposed to only be applied to the language it's specified for. The resolveMsg and resolveMsgAdvanced functions handle falling back to other languages if a certain key does not exist.

I've tested the effects of the changes manually by testing multiple configurations of different languages and translations, but I've not implemented an automated test.

The change is only breaking if the user of the library used a custom translation of an existing key in 1 language to translate that same key across multiple/all other languages.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Improved translation fallback on the login screens: theme-specific text now falls back to a secondary language when missing, reducing missing or incorrect text for users.

* **Refactor**
  * Streamlined message resolution to more reliably select theme and fallback translations.

* **Bug Fixes**
  * Fewer missing-translation issues and more consistent localized UI.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

BrDewit committed:
Fix for issue #875: Remove code that causes custom translations to be shown for languages that don't have a custom translation for a certain key

garronej commented:
Hello @BrDewit,  

Thank you for the well detailed PR.  

That being said I think this changes, while fixing the issue of overwriting the curstom translation of **an existing key** in only a specific language this change also cause important regression when defining custom message key.  

```ts
.withCustomTranslations({
    en: {
        myCustomMessage: "Hello world"
    }
})
```

With this PR applied, if the current language is `en`, `msgStr("myCustomMessage")` gives `"Hello World"` but if the current language is, let's say, `fr` then it gives "myCustomMessage".  

Defining custom, non standard message keys is the primary usecase of `withCustomTranslations`.  

What would be a reasonable approach is, if the key is a known one then do what you suggest otherwise keep the current behavior.

BrDewit commented:
Hi @garronej

That's a very reasonable remark, if they did not specify the custom message, you would indeed want to fall back to an existing translation as that's better than just having the key inserted in that position. Especially as that's probably a more common case than the one where a custom translation is given for an existing key. When I have time, I'll revise the changes to fix this regression.

BrDewit committed:
Revert "Fix for issue #875: Remove code that causes custom translations to be shown for languages that don't have a custom translation for a certain key"

This reverts commit 252cf5e3f46fa808958109a4cf5b4d0c4af1ff81.

BrDewit committed:
Merge branch 'keycloakify:main' into fix-875-custom-translation-bug

BrDewit committed:
Fix for issue #875: Separate the set of custom translations in a provided language from custom translations in the fallback language, and resolve to these sets in different places of the translation resolution logic

BrDewit commented:
@garronej excuse me for the delay, but I've finally made the change.

The translation logic will now first check in the custom translation of the provided language, then it will check the default set of the provided language, then it will check the custom translation of the fallback language, after which it will check the default set of the fallback language.

This way, my issue, where a custom translation in 1 language would override the default translation of that key in every language, in is fixed.
While the regression is also fixed, which would cause the key itself to be shown instead of a translation, when no custom translation of a non-default key was provided in the language, but it was in another.

The only breaking change is of course if someone used the behaviour I changed, as a feature, where they use one custom translation to override default keys for every language, but that behaviour is unintuitive. It seems reasonable that with `withCustomTranslations` you must provide a translation for the key in every language you wish to override.

BrDewit committed:
Merge branch 'keycloakify:main' into fix-875-custom-translation-bug
